src/lua/datetime.lua

```diff
@@ -0,0 +1,611 @@
+int     tnt_dt_dom          (dt_t dt);
+
+/* dt_parse_iso.h definitions */
+size_t tnt_dt_parse_iso_zone_lenient(const char *str, size_t len, int *offset);
+
+/* Tarantool functions - datetime.c */
+int    datetime_to_string(char *buf, int len, const struct datetime * date);
+size_t datetime_strftime(const struct datetime *date, char *buf, uint32_t len,
+                         const char *fmt);
+void   datetime_now(struct datetime *now);
+
+]]
+
+local builtin = ffi.C
+local math_modf = math.modf
+local math_floor = math.floor
+
+local SECS_PER_DAY     = 86400
+local NANOS_PER_SEC    = 1000000000
+
+-- Unix, January 1, 1970, Thursday
+local DAYS_EPOCH_OFFSET = 719163
+local SECS_EPOCH_OFFSET = DAYS_EPOCH_OFFSET * SECS_PER_DAY
+local TOSTRING_BUFSIZE = 48
+
+local datetime_t = ffi.typeof('struct datetime')
+
+local function is_datetime(o)
+    return ffi.istype(datetime_t, o)
+end
+
+local function check_date(o, message)
+    if not is_datetime(o) then
+        return error(("%s: expected datetime, but received %s"):
+                     format(message, o), 2)
+    end
+end
+
+local function check_table(o, message)
+    if type(o) ~= 'table' then
+        return error(("%s: expected table %s"):format(message, o), 2)
+    end
+end
+
+local function check_str(s, message)
+    if type(s) ~= 'string' then
+        return error(("%s: expected string, but received %s"):
+                     format(message, s), 2)
+    end
+end
+
+-- range may be of a form of pair {begin, end} or
+-- tuple {begin, end, negative}
+-- negative is a special value (so far) used for days only
+local function check_range(v, from, to, txt, extra)
+
+    if extra == v or (v >= from and v <= to) then
+        return
+    end
+
+    if extra == nil then
+        error(('value %d of %s is out of allowed range [%d, %d]'):
+              format(v, txt, from, to), 2)
+    else
+        error(('value %d of %s is out of allowed range [%d, %d..%d]'):
+              format(v, txt, extra, from, to), 2)
+    end
+end
+
+local function nyi(msg)
+    local text = 'Not yet implemented'
+    if msg ~= nil then
+        text = ("%s : '%s'"):format(text, msg)
+    end
+    error(text, 3)
+end
+
+-- convert from epoch related time to Rata Die related
+local function local_rd(secs)
+    return math_floor((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY)
+end
+
+-- convert UTC seconds to local seconds, adjusting by timezone
+local function local_secs(obj)
+    return obj.epoch + obj.tzoffset * 60
+end
+
+local function utc_secs(epoch, tzoffset)
+    return epoch - tzoffset * 60
+end
+
+-- get epoch seconds, shift to the local timezone
+-- adjust from 1970-related to 0000-related time
+-- then return dt in those coordinates (number of days
+-- since Rata Die date)
+local function local_dt(obj)
+    return builtin.tnt_dt_from_rdn(local_rd(local_secs(obj)))
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 then
+        secs = secs - 1
+        nsec = nsec + NANOS_PER_SEC
```

Comment:
can `nsec` be `-(2 * NANOS_PER_SEC)`?

Reply:
Oh, shit! I've fixed in the `tsafin/gh-5941-datetime-take2-wip` branch, but it was not there by some odd reason, while I was creating this pull-request.

Here is how it looks like now:
```lua
local function normalize_nsec(secs, nsec)
    if nsec < 0 or nsec >= NANOS_PER_SEC then
        secs = secs + math_floor(nsec / NANOS_PER_SEC)
        nsec = nsec % NANOS_PER_SEC
    end
    return secs, nsec
end
```

Reply:
Understood, as this problem became more apparent within interval arithmetics testings (where we could pass arbitrary `usec`/`nsec`/`msec`) I've committed fix as part of one of next patches 'datetime, lua: preliminary interval arithmetics'. This explains why it's not in current pull-request, as it was created from 2 other commits to the same branch. Will bring here as a small delta.

Reply:
Fixed by 11d912db3112038501876f7aaf5dcf33f3511277

Reply:
@Gerold103 ok, if you prefer this way - I'd update commits in-place. [I was asked by @unera to not force update branch, because some reviewers do not like that approach. If it's force update in pull request is the recommended way, then let's continue using it, result is much cleaner.]